addon/services/lazy-loader.js

```diff
@@ -55,7 +55,45 @@ export default Ember.Service.extend({
       return Ember.RSVP.resolve();
     }
 
-    return loadAssets(bundle.urls).then(()=>this.markBundleAsLoaded(bundle.name));
+    var urlsToLoad = this.getAssetUrlsForBundle(bundle);
+    return loadAssets(urlsToLoad).then(()=>this.markBundleAsLoaded(bundle.name));
+  },
+
+  /*
+
+    urlsToLoad: map of url -> true -- need O(1) set/get for adding items to the collection. passed recursively because bottom-up generation is O(n)
+    bundlesHandled: map of bundle -> true -- same as above, need O(1) collection management. used for cycle avoidance.
+
+  */
+  getAssetUrlsForBundle(bundle) {
+    var bundlesHandled = {};
```

Comment:
it's simpler to initialize it externally and then use the same object for all recursions.. make sense?